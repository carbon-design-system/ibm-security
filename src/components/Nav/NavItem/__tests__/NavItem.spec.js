/**
 * @file Navigation item tests.
 * @copyright IBM Security 2019
 */

import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { NavItem } from '../../../..';

describe('NavItem', () => {
  test('should have no Axe or DAP violations when rendering an internal link', async () => {
    const main = document.createElement('main');
    render(
      <ul role="menubar">
        <NavItem href="#">test nav item</NavItem>
      </ul>,
      {
        // DAP requires a landmark '<main>' in the DOM:
        container: document.body.appendChild(main),
      }
    );
    await expect(document.body).toHaveNoAxeViolations();
    await expect(document.body).toHaveNoDAPViolations(
      'NavItem as internal link'
    );
  });

  test('should have no Axe or DAP violations when rendering an external link', async () => {
    const main = document.createElement('main');
    render(
      <ul role="menubar">
        <NavItem href="https://www.ibm.com/">test nav item</NavItem>
      </ul>,
      {
        // DAP requires a landmark '<main>' in the DOM:
        container: document.body.appendChild(main),
      }
    );
    await expect(document.body).toHaveNoAxeViolations();
    await expect(document.body).toHaveNoDAPViolations(
      'NavItem as external link'
    );
  });

  test('should have no Axe or DAP violations when disabled', async () => {
    const main = document.createElement('main');
    const { debug } = render(
      <ul role="menubar">
        <NavItem href="#" disabled>
          test nav item
        </NavItem>
      </ul>,
      {
        // DAP requires a landmark '<main>' in the DOM:
        container: document.body.appendChild(main),
      }
    );
    debug();
    await expect(document.body).toHaveNoAxeViolations();
    await expect(document.body).toHaveNoDAPViolations(
      'NavItem that is disabled'
    );
  });

  test('should set tabindex to -1 when there are no children', () => {
    const { container } = render(<NavItem href="#" />);
    expect(container.firstElementChild).toHaveAttribute('tabindex', '-1');
  });

  test('should invoke click mock when clicked', () => {
    const onClickMock = jest.fn();
    const { getByText } = render(
      <NavItem href="#" onClick={onClickMock}>
        test item
      </NavItem>
    );
    userEvent.click(getByText(/test item/i));
    expect(onClickMock).toHaveBeenCalledTimes(1);
  });

  test('should invoke keypress mock when keypress detected', () => {
    const onKeyPressMock = jest.fn();
    const { getByText } = render(
      <NavItem href="#" onKeyPress={onKeyPressMock}>
        test item
      </NavItem>
    );
    fireEvent.keyPress(getByText(/test item/i), {
      key: 'Enter',
      code: 13,
      charCode: 13,
    });
    expect(onKeyPressMock).toHaveBeenCalledTimes(1);
  });

  test('should not be interactive when `disabled` is set to `true`', () => {
    const onClickMock = jest.fn();
    const { getByText } = render(
      <NavItem href="#" disabled onClick={onClickMock}>
        test item
      </NavItem>
    );
    userEvent.click(getByText(/test item/i));
    expect(onClickMock).toHaveBeenCalledTimes(0);
  });
});
