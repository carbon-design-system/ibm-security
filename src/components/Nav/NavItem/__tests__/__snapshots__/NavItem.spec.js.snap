// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavItem Rendering renders correctly 1`] = `
<li
  className="security--nav__list__item class security--nav__list__item--active"
  label=""
  onClick={[Function]}
  onKeyPress={[Function]}
  role="menuitem"
>
  <NavItemLink
    className="security--nav__list__item__link"
    element="a"
    href="#"
    id="security--nav__list__item"
    tabIndex={0}
  >
    Label
  </NavItemLink>
</li>
`;

exports[`NavItem Rendering renders the HTML of the node's subtree 1`] = `
<li
  class="security--nav__list__item class security--nav__list__item--active"
  label=""
  role="menuitem"
>
  <a
    class="security--nav__list__item__link"
    href="#"
    id="security--nav__list__item"
    tabindex="0"
  >
    Label
  </a>
</li>
`;
